### PR TITLE
Fix token doubling in DP async-scheduling overlap path (#395)

### LINKED ISSUE #395: Silent token duplication under async-scheduling + DP overlap (gpt-oss AIME score 0.13)
## Summary

With `data_parallel_size > 1` and async-scheduling enabled (the default on the tt platform), the DP overlap path in `step_dp_with_batch_queue` silently produces **duplicated sampled tokens at near-deterministic positions** (most commonly `<|end|><|end|>` and `<|start|>assistant<|start|>assistant…`). The token stream becomes invalid harmony, the parser bails into the partial-parse fallback, and the chat-completion response comes back with `content=null`. The server stays alive; throughput logs look fine; only the eval result reveals the damage. This is independent of #387 (the DP collective deadlock) — it surfaces even with PR #394 applied.

## Impact

On gpt-oss-120b, DP=4, AIME25 at `num_concurrent=16`:

| Config | Wall time | NULL content | Harmony parse errors | AIME25 score |
|---|---|---|---|---|
| `--no-async-scheduling` (sync path) | 35:18 | 0/30 | 0 | 0.7667 |
| async ON, with `viktorpus/fix-dp-zero-schedule` (PR #387) only | 51:21 | 18/30 | 12 | **0.1333** |
| async ON + the fix in #(this PR) | 37:29 | 0/30 | 0 | **0.80** |

PR #394 reduces but does not eliminate the corruption: with #394 alone, harmony errors collapse from 5 distinct patterns (`<|start|>assistantassistant`, `assistantassistant`, doubled `<|channel|>` / `final`, …) down to a single pattern (`<|end|><|end|>` only), but that one is still enough to give a 25 % null-content rate on a 16-request probe.

## Root cause

In `step_dp_with_batch_queue` (`plugins/vllm-tt-plugin/src/vllm_tt_plugin/engine.py`):

```python
finalize_before_submit = prev_handle is not None and (
    not global_has_requests
    or not prev_handle.overlap_ok
    or not current_overlap_ok
)

if finalize_before_submit:
    engine_core_outputs = _finalize_previous(prev_handle)

# Submit next step
if global_has_requests:
    next_handle = self.dp_gather_submit(scheduler_output, ...)

if not finalize_before_submit and prev_handle is not None:
    engine_core_outputs = _finalize_previous(prev_handle)
```

When `overlap_ok=True` on both prev and current, the engine submits the next step before draining the previous one. The submit eventually reaches `model_runner.build_model_input`, whose decode branch reads input tokens straight from the input batch state:

```python
input_positions = torch.from_numpy(input_batch.num_tokens[:num_reqs] - 1)
input_tokens = input_batch.token_ids_cpu_tensor[
    torch.arange(num_reqs), input_positions
].view(-1, 1)
```

That slot is only written inside `apply_and_build_runner_output`:

```python
self.input_batch.token_ids_cpu[rows, start_idxs] = sampled_token_ids_np
self.input_batch.num_tokens[rows] = end_idxs
```

…which runs from `apply_dp_execution_result`, which runs from `_finalize_previous`. In the overlap path that all happens **after** `dp_gather_submit` has already read the (stale) `input_batch`. The device therefore re-runs the previous step's input and, for near-deterministic positions, samples the previous step's output token a second time.

The sync `step` path is unaffected because `_execute_model_dp_gather` calls `dp_gather_finalize` immediately after `dp_gather_submit` in the same call — every iteration's `apply_dp_execution_result` has finished before the next iteration's `step` is even entered. This is why `--no-async-scheduling` produces 0 doubled tokens.

## Token-level evidence

Three failing requests on `gpt-oss-120b` async-ON + PR #394, captured with extended logging on `harmony_utils.parse_output_into_messages`:

```
last30_ids = [..., 13, 13, 200007, 200007]
decoded     = '...provide solution.<|end|><|end|>'

last30_ids = [..., 364, 200007, 200007]
decoded     = '...steps.\n\n<|end|><|end|>'

last30_ids = [..., 25, 13, 200007, 200007]
decoded     = '...<|end|><|end|>'
```

The pattern is consistent: the last "valid" token (`13` = `.`, `364` = ` "`, etc.) is followed by `200007` (`<|end|>`), and then `200007` appears again where harmony expects `200006` (`<|start|>`).

## Why duplication is identical even with `temperature=1.0`

The doubled tokens are byte-for-byte identical (e.g. content tokens `13, 13` and `30854, 30854`), not just "two random samples from the same logits". The cleanest explanation is that the next step's compute receives the *same input* as the previous step (because `input_batch.token_ids_cpu` hadn't been updated), and the sample falls on the same near-deterministic position. Stochastic content tokens are usually masked by RNG advancing within a single decode call, but boundary tokens with probability ≈ 1.0 always reproduce identically.

## Fix

Force `finalize_before_submit = prev_handle is not None`. This makes the apply-then-submit ordering hold every iteration: `_finalize_previous` (which calls `apply_dp_execution_result` and updates `input_batch`) always runs before the next `dp_gather_submit`, so the next step's `build_model_input` sees the previous step's sampled tokens. Posted as PR for this fix.

Cost: the host-side finalize work (`apply_dp_execution_result` + `update_from_output`) is no longer hidden behind the next device kick-off. On the AIME25 workload above the cost is ~6 % vs the sync path. The proper structural fix preserves the overlap by splitting `apply_and_build_runner_output` into a fast "apply tokens" half (run before submit) and a slower "build runner output" half (run after submit); that is worth doing in a follow-up but the diff is non-trivial.

## Reproduction

Server (gpt-oss-120b, DP=4 on a Wormhole galaxy):

```bash
cd $TT_METAL_HOME && source python_env/bin/activate
export HF_MODEL=/path/to/gpt-oss-120b TT_CACHE_PATH=/path/to/tt_cache
export MESH_DEVICE="(4, 8)" TT_MM_THROTTLE_PERF=2 VLLM_RPC_TIMEOUT=900000
cd $VLLM_HOME
python plugins/vllm-tt-plugin/examples/server_example_tt.py \
  --model openai/gpt-oss-120b --max_num_seqs 32 --data_parallel_size 4 \
  --plugin-config '{"tt":{"fabric_config":"FABRIC_1D_RING","sample_on_device_mode":"decode_only"}}'
```

(Notice the absence of `--no-async-scheduling`. async is the default.)

Client:

```bash
lm_eval --tasks aime25 --model local-chat-completions \
  --model_args model=openai/gpt-oss-120b,base_url=http://127.0.0.1:8000/v1/chat/completions,tokenizer_backend=huggingface,timeout=7200,num_concurrent=16 \
  --gen_kwargs reasoning_effort=high,do_sample=true,temperature=1.0,max_gen_toks=65536 \
  --seed 42 --num_fewshot 0 --batch_size 1 --apply_chat_template
```

Expected on broken master: ~18/30 chat completions return `content=null` ("API returned null content" lm-eval warning), final AIME25 score 0.13 ± 0.06. With the fix: 0 null content responses, score recovers to 0.77–0.80.

## Related

* #387 – DP collective deadlock; PR #394 (`viktorpus/fix-dp-zero-schedule`) fixes that one. That fix is necessary but not sufficient; the bug described here remains after #394.
* #386 – harmony parser robustness; the `Harmony parser error … returning partial parse` log line in this report comes from the partial-parse fallback added there.

### PR DESCRIPTION
**Pull request by sraizada-tt**
## Summary

Fixes #395 — silent token duplication under async-scheduling + DP overlap. With async-scheduling on (default on tt platform) and `data_parallel_size > 1`, `step_dp_with_batch_queue`'s overlap path submitted the next step before applying the previous step's sampled tokens to `input_batch`. The next step's `build_model_input` then pulled stale input tokens, the device re-sampled the previous near-deterministic position, and chat completions silently returned `content=null` from harmony parse failures.

This PR forces `finalize_before_submit = prev_handle is not None` every iteration, restoring apply-then-submit ordering.

Root-cause walkthrough and token-level evidence are in #395.

## Test plan

Measured on `gpt-oss-120b`, DP=4 on a Wormhole 4x8 galaxy, `vllm + viktorpus/fix-dp-zero-schedule` base (PR #394 applied) + this PR. `lm_eval --tasks aime25 --num_concurrent 16 --gen_kwargs reasoning_effort=high,do_sample=true,temperature=1.0,max_gen_toks=65536 --seed 42`:

| Config | Wall time | NULL content | Harmony parse errors | AIME25 score |
|---|---|---|---|---|
| `--no-async-scheduling` baseline | 35:18 | 0/30 | 0 | 0.7667 |
| async ON, before this PR (PR #394 only) | 51:21 | 18/30 | 12 | 0.1333 |
| async ON, after this PR | **37:29** | **0/30** | **0** | **0.80** |

Also ran a 16-request concurrent probe of mixed reasoning prompts directly against the chat endpoint:

| Config | NULL content | Harmony parse errors |
|---|---|---|
| async ON, before this PR (PR #394 only) | 4/16 | 4 |
| async ON, after this PR | 0/16 | 0 |

- [x] AIME25 30/30, async ON, no NULL responses, no harmony errors
- [x] AIME25 score within noise of `--no-async-scheduling` baseline
- [x] No regression on sync path (`step` is unaffected — only the async `step_dp_with_batch_queue` path changes)
- [x] Hang from #387 still does not reproduce on top of #394 (`tt-metal sraizada/gpt-oss-aime-release` @ `f7e2f8b499a`, 4x8 Galaxy, GWH01)

## Performance note

The host-side finalize work (`apply_dp_execution_result` + `update_from_output`) is no longer hidden behind the next device kick-off. Cost vs the sync path on the AIME25 workload above is ~6%. The async path is still 27% faster than its broken-overlap predecessor because rejected null responses no longer waste decode time.

A structural fix that preserves overlap is possible by splitting `apply_and_build_runner_output` into a fast "apply tokens to `input_batch`" half (run before submit) and a slower "build runner output" half (run after submit). That diff is non-trivial and touches the worker / model_runner / engine layers, so I'd suggest landing this minimal fix first and doing the structural one as a follow-up if the ~6% perf delta is felt.

### CODE REVIEW

**Comment by sraizada-tt** [comment]
Verified on standalone PR branch (`dev` + this commit, **without** PR #394 applied):

```
lm_eval --tasks aime25 --limit 6 ... --num_concurrent 16 ...
```

| Metric | Value |
|---|---|
| Wall time | 8:15 |
| AIME25 score | 1.0 (6/6) |
| NULL content responses | 0/6 |
| Harmony parse errors | 0 |
| Server errors | 0 |

Confirms the fix lands cleanly on `dev` and doesn't depend on PR #394.

**Review by viktorpusTT — APPROVED** [review]
Expect to see performance drop. Approve for now. I have WiP to get back to full async overlap, but need to split forward from sampling in model API.

File: setup.py [hunk 1]
```diff
@@ -817,6 +817,10 @@ def _is_xpu() -> bool:
     return VLLM_TARGET_DEVICE == "xpu"
 
 
+def _is_tt() -> bool:
```
Comment (viktorpusTT) [inline setup.py]: We want to avoid TT-specific code outside of plugin. To install VLLM for TT, VLLM_TARGET_DEVICE must be set to "empty", as documented at https://github.com/tenstorrent/vllm/tree/dev/plugins/vllm-tt-plugin

**Review by viktorpusTT — CHANGES_REQUESTED** [review]

**Comment by sraizada-tt** [comment]
Additional context for reviewers: while validating this PR with `sample_on_device_mode=all`, a separate prefill-side issue surfaced (different code path, different symptoms). Tracked at [tenstorrent/tt-metal#44746](https://github.com/tenstorrent/tt-metal/issues/44746). It does **not** affect this PR's correctness for the `decode_only` config used in the test plan; mentioning it so it doesn't get conflated.

Suggested deployment recommendation alongside this PR: `--plugin-config '{"tt":{...,"sample_on_device_mode":"decode_only"}}'` until the tt-metal fix lands.

**Review by viktorpusTT — APPROVED** [review]
